src/ndb.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+#include "ndb.h"
+
+ndb_file_version_t ndb_file_version = 0x00000001;
+
+/* Period of repeating changes write operation, milliseconds */
+#define NDB_WRITE_PERIOD 500
+
+typedef Mutex ndb_mutex_t;
+typedef Thread *ndb_thread_t;
+
+static ndb_thread_t ndb_thread;
+static WORKING_AREA(ndb_thread_wa, 128);
+static msg_t nv_write_thread(void*);
+static ndb_mutex_t data_access;
+
+static enum ndb_op_code ndb_open_file(ndb_file_t *file, ndb_file_version_t *v);
+static enum ndb_op_code ndb_read(ndb_file_t *f, void *first_element, size_t s);
+
+#define OPTF __attribute__ ((weak, alias ("ndb_not_implemented")))
+
+enum ndb_op_code ndb_not_implemented()
+{
+    return NDB_ERR_UNSUPPORTED;
+}
+
+enum ndb_op_code ndb_p1_init() OPTF;
```

Comment:
I think the linked list of structures for each section was good for the `init()` and `nvm_write()` interfaces. You'll always be iterating through all sections when these operations run, anyway.